### PR TITLE
Code comments (issue #303).

### PR DESCRIPTION
**Pull request by aquental**
Code comments for all functions in Starknet.go library.
Issue [#303](https://github.com/NethermindEth/starknet.go/issues/303)

### CODE REVIEW

File: account/account_test.go [hunk 1]
```diff
@@ -721,6 +833,25 @@ func TestWaitForTransactionReceipt(t *testing.T) {
...
+// After that, it creates a new RPC client and provider.
+// Then, it creates a new account using the provider and the account address, public key, and keystore.
+// The function reads the content of the "./tests/hello_starknet_compiled.sierra.json" file and unmarshals it into a ContractClass struct.
+// It calculates the class hash using the hash.ClassHash function.
+// The function then reads the content of the "./tests/hello_starknet_compiled.sierra.json" file again and unmarshals it into a CasmClass struct.
+// It calculates the compiled class hash using the hash.CompiledClassHash function.
+// Next, it retrieves the nonce using the account's Nonce method.
+// The function creates a new DeclareTxnV2 struct and sets its fields.
+// It signs the declare transaction using the SignDeclareTransaction method of the account.
+// Finally, it adds the declare transaction using the AddDeclareTransaction method of the account and checks the response.
```
Comment (rianhughes) [inline account/account_test.go]: I don't think we need this level of detail, the reader can just look at the code to see what's being done.

File: account/account_test.go [hunk 2]
```diff
@@ -721,6 +833,25 @@ func TestWaitForTransactionReceipt(t *testing.T) {
...
+//
+// No parameters.
+// No return types.
```
Comment (rianhughes) [inline account/account_test.go]: Could we consistently use the same format, eg
Parameters : None
Returns: None

File: account/account_test.go [hunk 3]
```diff
@@ -792,6 +923,10 @@ func TestAddDeclareTxn(t *testing.T) {
 	}
 }
 
+// newDevnet creates a new devnet with the given URL.
+//
+// It takes a testing object `t` and a string `url` as parameters.
+// It returns a pointer to a `devnet.DevNet` object, a slice of `devnet.TestAccount`, and an error.
```
Comment (rianhughes) [inline account/account_test.go]: Same here,
Parameters:..
Returns:..

File: account/keystore.go [hunk 4]
```diff
@@ -63,7 +92,18 @@ func (ks *MemKeystore) Sign(ctx context.Context, id string, msgHash *big.Int) (*
...
+//
+// It takes the following parameters:
+// - ctx: the context.Context object for cancellation and timeouts.
+// - msgHash: the message hash to be signed as a *big.Int.
+// - key: the private key as a *big.Int.
+//
+// It returns the following values:
+// - x: the X coordinate of the signature point as a *big.Int.
+// - y: the Y coordinate of the signature point as a *big.Int.
+// - err: an error object if any error occurred during the signing process.
```
Comment (rianhughes) [inline account/keystore.go]: This format is nice!

File: contracts/contracts_test.go [hunk 5]
```diff
@@ -22,6 +26,16 @@ func TestUnmarshalContractClass(t *testing.T) {
 	assert.Equal(t, class.SierraProgram[1].String(), "0x3")
 }
 
+// TestUnmarshalCasmClass tests the UnmarshalCasmClass function.
+//
+// It checks if the UnmarshalCasmClass function correctly unmarshals the contents of the casmClass file.
+// The function takes a file path as a parameter and returns the unmarshalled CasmClass object and an error.
+// The function asserts that the Prime field of the CasmClass object is equal to "0x800000000000011000000000000000000000000000000000000000000000001",
+// the Version field is equal to "2.1.0", the Selector field of the first External entry point is equal to "0x362398bec32bc0ebb411203221a35a0301193a96f317ebe5e40be9f60d15320",
+// the Offset field of the second External entry point is equal to 130, and the first Builtin of the second External entry point is equal to "range_check".
```
Comment (rianhughes) [inline contracts/contracts_test.go]: Imo, I don't think we need this level of detail

**Comment by aquental** [comment]
I reviewed all files.
Now they are using the same format.

File: account/account.go [hunk 6]
```diff
@@ -213,6 +274,35 @@ func (account *Account) TransactionHashInvoke(tx rpc.InvokeTxnType) (*felt.Felt,
...
+//       - It calls the `hash.CalculateTransactionHashCommon` function with the following parameters:
+//           - `PREFIX_DECLARE`
+//           - `txnVersionFelt`
+//           - `txn.SenderAddress`
+//           - `&felt.Zero`
+//           - `calldataHash`
+//           - `txn.MaxFee`
+//           - `account.ChainId`
+//           - A slice containing `txn.Nonce`
+//           - For `rpc.DeclareTxnV2`, it also includes `txn.CompiledClassHash` in the slice of parameters.
```
Comment (cicr99) [inline account/account.go]: This is ok but I think it's too detailed. I would just leave it on:
```
// Can be one of the following types:
//   - `rpc.DeclareTxnV0`
//      - for `rpc.DeclareTxnV0`, the function returns an error `ErrTxnVersionUnSupported`.
//   - `rpc.DeclareTxnV1`
//   - `rpc.DeclareTxnV2`
```

File: account/account.go [hunk 7]
```diff
@@ -316,103 +423,329 @@ func (account *Account) WaitForTransactionReceipt(ctx context.Context, transacti
 	}
 }
 
+// AddInvokeTransaction generates an invoke transaction and adds it to the account's provider.
+//
+// Parameters:
+// ctx - the context.Context object for the transaction.
+// invokeTx - the invoke transaction to be added.
```
Comment (cicr99) [inline account/account.go]: The format is different here
```suggestion
// - ctx: the context.Context object for the transaction.
// - invokeTx: the invoke transaction to be added.
```

File: account/account.go [hunk 8]
```diff
@@ -316,103 +423,329 @@ func (account *Account) WaitForTransactionReceipt(ctx context.Context, transacti
...
+// - error: an error if any.
 func (account *Account) AddInvokeTransaction(ctx context.Context, invokeTx rpc.InvokeTxnV1) (*rpc.AddInvokeTransactionResponse, error) {
 	return account.provider.AddInvokeTransaction(ctx, invokeTx)
 }
 
+// AddDeclareTransaction adds a declare transaction to the account.
+//
+// Parameters:
+// ctx: The context.Context for the request.
+// declareTransaction: The input for adding a declare transaction.
```
Comment (cicr99) [inline account/account.go]:
```suggestion
// - ctx: The context.Context for the request.
// - declareTransaction: The input for adding a declare transaction.
```

File: account/account.go [hunk 9]
```diff
@@ -316,103 +423,329 @@ func (account *Account) WaitForTransactionReceipt(ctx context.Context, transacti
...
+// - error: an error, if any
 func (account *Account) AddDeclareTransaction(ctx context.Context, declareTransaction rpc.AddDeclareTxnInput) (*rpc.AddDeclareTransactionResponse, error) {
 	return account.provider.AddDeclareTransaction(ctx, declareTransaction)
 }
 
+// AddDeployAccountTransaction adds a deploy account transaction to the account.
+//
+// Parameters:
+// ctx: The context.Context object for the function.
+// deployAccountTransaction: The rpc.DeployAccountTxn object representing the deploy account transaction.
```
Comment (cicr99) [inline account/account.go]:
```suggestion
// - ctx: The context.Context object for the function.
// - deployAccountTransaction: The rpc.DeployAccountTxn object representing the deploy account transaction.
```

File: account/account.go [hunk 10]
```diff
@@ -316,103 +423,329 @@ func (account *Account) WaitForTransactionReceipt(ctx context.Context, transacti
...
+// - *rpc.AddDeployAccountTransactionResponse: a pointer to rpc.AddDeployAccountTransactionResponse
+// - error: an error if any
 func (account *Account) AddDeployAccountTransaction(ctx context.Context, deployAccountTransaction rpc.DeployAccountTxn) (*rpc.AddDeployAccountTransactionResponse, error) {
 	return account.provider.AddDeployAccountTransaction(ctx, deployAccountTransaction)
 }
 
+// BlockHashAndNumber returns the block hash and number for the account.
+//
+// Parameters:
+// ctx - The context in which the function is called.
```
Comment (cicr99) [inline account/account.go]:
```suggestion
// - ctx: The context in which the function is called.
```

File: account/account.go [hunk 11]
```diff
@@ -316,103 +423,329 @@ func (account *Account) WaitForTransactionReceipt(ctx context.Context, transacti
...
 }
 
+// BlockTransactionCount returns the number of transactions in a block.
+//
+// Parameters:
+// - ctx: The context.Context object for the function.
+// - blockID: The rpc.BlockID object representing the block.
+// Returns:
+// - uint64: the number of transactions in the block
+//  - error: an error, if any
```
Comment (cicr99) [inline account/account.go:476]:
```suggestion
// - error: an error, if any
```

File: account/account.go [hunk 12]
```diff
@@ -316,103 +423,329 @@ func (account *Account) WaitForTransactionReceipt(ctx context.Context, transacti
...
 func (account *Account) BlockTransactionCount(ctx context.Context, blockID rpc.BlockID) (uint64, error) {
 	return account.provider.BlockTransactionCount(ctx, blockID)
 }
 
+// BlockWithTxHashes retrieves a block with transaction hashes.
+//
+// Parameters:
+// ctx - the context.Context object for the request.
+// blockID - the rpc.BlockID object specifying the block to retrieve.
```
Comment (cicr99) [inline account/account.go]:
```suggestion
// - ctx: the context.Context object for the request.
// - blockID: the rpc.BlockID object specifying the block to retrieve.
```

File: account/account.go [hunk 13]
```diff
@@ -316,103 +423,329 @@ func (account *Account) WaitForTransactionReceipt(ctx context.Context, transacti
...
+// - interface{}: An interface{}
+// - error: An error
 func (account *Account) BlockWithTxs(ctx context.Context, blockID rpc.BlockID) (interface{}, error) {
 	return account.provider.BlockWithTxs(ctx, blockID)
 }
 
+// Call is a function that performs a function call on an Account.
+//
+// Parameters:
+// It takes in a context.Context object, a rpc.FunctionCall object, and a rpc.BlockID object as parameters.
```
Comment (cicr99) [inline account/account.go]: Could these parameters be formatted the same way as the others?

File: account/account.go [hunk 14]
```diff
@@ -316,103 +423,329 @@ func (account *Account) WaitForTransactionReceipt(ctx context.Context, transacti
...
+// Parameters:
+// - ctx: the context.Context object for the function.
+// Returns:
+//   - string: the chain ID.
+//   - error: any error encountered while retrieving the chain ID.
 func (account *Account) ChainID(ctx context.Context) (string, error) {
 	return account.provider.ChainID(ctx)
 }
+
+// Class is a method that calls the `Class` method of the `provider` field of the `account` struct.
```
Comment (cicr99) [inline account/account.go:529]: I would change this to:
"Class gets the contract class definition associated with the given hash."

File: account/account.go [hunk 15]
```diff
@@ -316,103 +423,329 @@ func (account *Account) WaitForTransactionReceipt(ctx context.Context, transacti
...
+//
+// Parameters:
+// - ctx: The context.Context
+// - blockID: The rpc.BlockID 
+// - classHash: The `*felt.Felt`
+// Returns:
+// - *rpc.ClassOutput: a `rpc.ClassOutput`
```
Comment (cicr99) [inline account/account.go]: The class output could be a DeprecatedContractClass or just a Contract class depending on the contract version

File: account/account.go [hunk 16]
```diff
@@ -316,103 +423,329 @@ func (account *Account) WaitForTransactionReceipt(ctx context.Context, transacti
...
 }
+
+// ClassAt retrieves the class at the specified block ID and contract address.
+// Parameters:
+// - ctx: The context.Context object for the function.
+// - blockID: The rpc.BlockID object representing the block ID.
+// - contractAddress: The felt.Felt object representing the contract address.
+// 
+// Returns:
+// - *rpc.ClassOutput: The rpc.ClassOutput object
```
Comment (cicr99) [inline account/account.go]: Same here about the ClassOutput

File: account/account.go [hunk 17]
```diff
@@ -316,103 +423,329 @@ func (account *Account) WaitForTransactionReceipt(ctx context.Context, transacti
...
 func (account *Account) ClassAt(ctx context.Context, blockID rpc.BlockID, contractAddress *felt.Felt) (rpc.ClassOutput, error) {
 	return account.provider.ClassAt(ctx, blockID, contractAddress)
 }
 
+// ClassHashAt returns the class hash at the given block ID for the specified contract address.
+//
+// Parameters:
+// ctx - The context to use for the function call.
+// blockID - The ID of the block.
+// contractAddress - The address of the contract to get the class hash for.
```
Comment (cicr99) [inline account/account.go]:
```suggestion
// - ctx: The context to use for the function call.
// - blockID: The ID of the block.
// - contractAddress: The address of the contract to get the class hash for.
```

File: account/account.go [hunk 18]
```diff
@@ -316,103 +423,329 @@ func (account *Account) WaitForTransactionReceipt(ctx context.Context, transacti
...
 func (account *Account) EstimateFee(ctx context.Context, requests []rpc.EstimateFeeInput, blockID rpc.BlockID) ([]rpc.FeeEstimate, error) {
 	return account.provider.EstimateFee(ctx, requests, blockID)
 }
+
+// EstimateMessageFee estimates the fee for a given message in the context of an account.
+//
+// Parameters:
+// ctx - The context.Context object for the function.
+// msg - The rpc.MsgFromL1 object representing the message.
+// blockID - The rpc.BlockID object representing the block ID.
```
Comment (cicr99) [inline account/account.go]:
```suggestion
// - ctx: The context.Context object for the function.
// - msg: The rpc.MsgFromL1 object representing the message.
// - blockID: The rpc.BlockID object representing the block ID.
```

File: account/account.go [hunk 19]
```diff
@@ -316,103 +423,329 @@ func (account *Account) WaitForTransactionReceipt(ctx context.Context, transacti
...
 func (account *Account) Events(ctx context.Context, input rpc.EventsInput) (*rpc.EventChunk, error) {
 	return account.provider.Events(ctx, input)
 }
+
+// Nonce returns the nonce for the specified account and contract address.
+//
+// Parameters:
+// - ctx: The context.Context object for the function.
+// - blockID: The rpc.BlockID object for the function.
+// - contractAddress: The felt.Felt object for the function.
```
Comment (cicr99) [inline account/account.go]: The address of the contract whose nonce we're seeking

File: account/account.go [hunk 20]
```diff
@@ -316,103 +423,329 @@ func (account *Account) WaitForTransactionReceipt(ctx context.Context, transacti
...
+// - *string: a string pointer
+// - error: an error
 func (account *Account) Nonce(ctx context.Context, blockID rpc.BlockID, contractAddress *felt.Felt) (*string, error) {
 	return account.provider.Nonce(ctx, blockID, contractAddress)
 }
 
+// SimulateTransactions simulates transactions using the provided context
+// Parameters: 
+// - ctx: The context.Context object
+// - blockID: The rpc.BlockID object
```
Comment (cicr99) [inline account/account.go]:
```suggestion
// - blockID: The rpc.BlockID object for the block referencing the state or call the transactions are on
```

File: account/account.go [hunk 21]
```diff
@@ -316,103 +423,329 @@ func (account *Account) WaitForTransactionReceipt(ctx context.Context, transacti
...
+// - txns: The slice of rpc.Transaction objects
```
Comment (cicr99) [inline account/account.go]:
```suggestion
// - txns: The slice of rpc.Transaction objects representing the transactions to simulate
```

**Comment by aquental** [comment]
Please, check if my PR is ok.
Thanks

**Review by cicr99 — APPROVED** [review]

**Pull request merged**